src/lib.rs

```diff
@@ -727,15 +727,216 @@ impl CertificateParams {
+			}
+			if key_usage.decipher_only() {
+				key_usages.push(KeyUsagePurpose::DecipherOnly);
+			}
+		}
+		Ok(key_usages)
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_extended_key_usages(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<Vec<ExtendedKeyUsagePurpose>, RcgenError> {
+		let extended_key_usage = x509.extended_key_usage()
+			.or(Err(RcgenError::CouldNotParseCertificate))?
+			.map(|ext| ext.value);
+
+		let mut extended_key_usages = Vec::new();
+		if let Some(extended_key_usage) = extended_key_usage {
+			if extended_key_usage.any {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::Any);
+			}
+			if extended_key_usage.server_auth {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::ServerAuth);
+			}
+			if extended_key_usage.client_auth {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::ClientAuth);
+			}
+			if extended_key_usage.code_signing {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::CodeSigning);
+			}
+			if extended_key_usage.email_protection {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::EmailProtection);
+			}
+			if extended_key_usage.time_stamping {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::TimeStamping);
+			}
+			if extended_key_usage.ocsp_signing {
+				extended_key_usages.push(ExtendedKeyUsagePurpose::OcspSigning);
+			}
+		}
+		Ok(extended_key_usages)
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_name_constraints(x509 :&x509_parser::certificate::X509Certificate<'_>) -> Result<Option<NameConstraints>, RcgenError> {
+		let constraints = x509.name_constraints()
+			.or(Err(RcgenError::CouldNotParseCertificate))?
+			.map(|ext| ext.value);
+
+		if let Some(constraints) = constraints {
+			let permitted_subtrees = if let Some(permitted) = &constraints.permitted_subtrees {
+				Self::convert_x509_general_subtrees(&permitted)?
+			} else {
+				Vec::new()
+			};
+
+			let excluded_subtrees = if let Some(excluded) = &constraints.excluded_subtrees {
+				Self::convert_x509_general_subtrees(&excluded)?
+			} else {
+				Vec::new()
+			};
+
+			let name_constraints = NameConstraints { permitted_subtrees, excluded_subtrees };
+
+			Ok(Some(name_constraints))
+		} else {
+			Ok(None)
+		}
+	}
+	#[cfg(feature = "x509-parser")]
+	fn convert_x509_general_subtrees(subtrees :&[x509_parser::extensions::GeneralSubtree<'_>]) -> Result<Vec<GeneralSubtree>, RcgenError> {
+		use x509_parser::extensions::GeneralName;
+
+		let mut result = Vec::new();
+		for subtree in subtrees {
+			let subtree = match &subtree.base {
+				GeneralName::RFC822Name(s) => GeneralSubtree::Rfc822Name(s.to_string()),
+				GeneralName::DNSName(s) => GeneralSubtree::DnsName(s.to_string()),
+				GeneralName::DirectoryName(n) => GeneralSubtree::DirectoryName(DistinguishedName::from_name(&n)?),
+				GeneralName::IPAddress(bytes) if bytes.len() == 8 => {
+					let addr: [u8; 4] = bytes[..4].try_into().unwrap();
+					let mask: [u8; 4] = bytes[4..].try_into().unwrap();
```

Comment:
Should be possible to write this in a way that avoids `unwrap`

Reply:
`unwrap` is fine here: if it gets triggered, then something bad happened to the `== 8` above, e.g. in a refactor, which would indicate a bug. It likely gets removed by the optimizer anyways. I prefer something with unwrap over something longer or something that uses unsafe.

Reply:
See my proposed implementation in #101